### ISSUE TITLE
Sporadic build failures after upgrading to 0.15.2: "java.util.concurrent.RejectedExecutionException: event executor terminated"

### ISSUE THREAD

**Issue by roblg** [bazelbuild/bazel#5711]
### Description of the problem:

After upgrading from bazel-0.13.x to bazel 0.15.2, we're getting sporadic build failures where the build appears to get all the way to the end successfully, and then fails. I don't know if this is specific to bazel 0.15.2, but it appears to at least occur more frequently now than it used to.

It seems relevant that we're using the http-based action cache (SimpleBlobStoreActionCache+HttpBlobStore).

The failure stack trace is:

```
15:48:59 java.util.concurrent.RejectedExecutionException: event executor terminated
15:48:59 	at io.netty.util.concurrent.SingleThreadEventExecutor.reject(SingleThreadEventExecutor.java:821)
15:48:59 	at io.netty.util.concurrent.SingleThreadEventExecutor.offerTask(SingleThreadEventExecutor.java:327)
15:48:59 	at io.netty.util.concurrent.SingleThreadEventExecutor.addTask(SingleThreadEventExecutor.java:320)
15:48:59 	at io.netty.util.concurrent.SingleThreadEventExecutor.execute(SingleThreadEventExecutor.java:746)
15:48:59 	at io.netty.channel.pool.FixedChannelPool.close(FixedChannelPool.java:440)
15:48:59 	at com.google.devtools.build.lib.remote.blobstore.http.HttpBlobStore.close(HttpBlobStore.java:473)
15:48:59 	at com.google.devtools.build.lib.remote.SimpleBlobStoreActionCache.close(SimpleBlobStoreActionCache.java:224)
15:48:59 	at com.google.devtools.build.lib.remote.RemoteActionContextProvider.executionPhaseEnding(RemoteActionContextProvider.java:116)
15:48:59 	at com.google.devtools.build.lib.buildtool.ExecutionTool.shutdown(ExecutionTool.java:203)
15:48:59 	at com.google.devtools.build.lib.buildtool.BuildTool.buildTargets(BuildTool.java:284)
15:48:59 	at com.google.devtools.build.lib.buildtool.BuildTool.processRequest(BuildTool.java:350)
15:48:59 	at com.google.devtools.build.lib.runtime.commands.TestCommand.doTest(TestCommand.java:126)
15:48:59 	at com.google.devtools.build.lib.runtime.commands.TestCommand.exec(TestCommand.java:106)
15:48:59 	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.execExclusively(BlazeCommandDispatcher.java:472)
15:48:59 	at com.google.devtools.build.lib.runtime.BlazeCommandDispatcher.exec(BlazeCommandDispatcher.java:199)
15:48:59 	at com.google.devtools.build.lib.server.GrpcServerImpl.executeCommand(GrpcServerImpl.java:855)
15:48:59 	at com.google.devtools.build.lib.server.GrpcServerImpl.access$2100(GrpcServerImpl.java:111)
15:48:59 	at com.google.devtools.build.lib.server.GrpcServerImpl$2.lambda$run$0(GrpcServerImpl.java:924)
15:48:59 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
15:48:59 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
15:48:59 	at java.lang.Thread.run(Thread.java:748)
```

Based on some initial investigation, it seems like there are two code paths per build that trigger `ActionContextProvider.executionPhaseEnding` to be called -- one in `ExecutionTool.executeBuild`, and one when `ExecutionTool.shutdown` is called from `BuildTool`.

It appears that this causes SimpleBlobStoreActionCache.close() to be called twice, which results in a double-close on the underlying netty worker pool. I don't know why it only causes the build to fail _sometimes_.

**Possible easy fix?** In the "other" cache implementation, [`ByteStreamUploader.shutdown()`](https://github.com/bazelbuild/bazel/blob/0.15.2/src/main/java/com/google/devtools/build/lib/remote/ByteStreamUploader.java#L167-L171), which is called by [`GrpcRemoteCache.close()`](https://github.com/bazelbuild/bazel/blob/0.15.2/src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteCache.java#L112-L114) has a guard to handle being called multiple times; perhaps [HttpBlobStore.close()](https://github.com/bazelbuild/bazel/blob/0.15.2/src/main/java/com/google/devtools/build/lib/remote/blobstore/http/HttpBlobStore.java#L472-L475) should have one as well?

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

We don't have a particularly reliable repro. It's failing ~5% of our continuous integration builds.

### What operating system are you running Bazel on?

We see the failures on our Linux CI builders specifically (CentOS 7).

### What's the output of `bazel info release`?

release 0.15.2

###  Have you found anything relevant by searching the web?

Couldn't find any mention of this error, which I find pretty surprising, not going to lie. :)

### Any other information, logs, or outputs that you want to share?

I'll try to roll an internal build based on 0.15.2 that includes a simple guard in HttpBlobStore.close() and see if that addresses our issues.

**Comment by jin** [comment]
Assigning this to @buchgr, since it looks related to remote caching.

**Comment by crorvick** [comment]
I'm seeing the same problem with 0.15.1, which is probably expected since it is identical to 0.15.2.

**Comment by buchgr** [comment]
@roblg I think it doesn't. Looking at this code, `executor.inEventLoop()` will be false and so Netty will try to schedule the close task on the executor which has been shutdown [1]. So I think your suggested fix is correct. Do you want to send a PR?

Additionally, I ll investigate why executionPhaseEnding() is called multiple times. 

[1] https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java#L449

**Comment by roblg** [comment]
I can take a stab at a PR, yep. I just want to make sure we're looking at the same things first though. :)

Based on bazel/third_party/netty, it looks like bazel depends on netty 4.1.17, which has slightly different code than what you linked there ([Netty 4.1.17 FixedChannelPool](https://github.com/netty/netty/blob/netty-4.1.17.Final/transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java#L439)). I think your explanation still makes sense though. In [FixedChannelPool.java#443-445](https://github.com/netty/netty/blob/netty-4.1.17.Final/transport/src/main/java/io/netty/channel/pool/FixedChannelPool.java#L443-L445), the scheduled `Runnable` handles the close, but that wouldn't prevent the `executor` from blowing up, which is actually what we see.

Thank you for talking me out of talking myself out of that fix. 😌

**Comment by buchgr** [comment]
awesome! Please send me a PR and we can cherry pick it into 0.17.0!

**Comment by roblg** [comment]
@buchgr I had trouble reproing this issue even with a forced, minimal test case (in HttpBlobStoreTest), so I started digging around some more.

It looks like this is actually a netty bug that's fixed as of ~4.1.26: https://github.com/netty/netty/issues/7901. The fix makes sure `close()` doesn't return until it's finished. What're the odds of upgrading bazel's netty dependency for 0.17.x? :)

I'll still submit the change to prevent HttpBlobStore from doing work in `close()` more than once, since that also seems desirable.

**Comment by yiming9** [comment]
On release 0.15.0, we were also hit by this bug several times (about once or twice per day), particularly when a lot of data were written into the remote cache (e.g., compiling statically-linked binaries with debug symbols). It would be great if we have a patch on this issue. Could we merge https://github.com/bazelbuild/bazel/pull/5725? Thanks a lot!

**Comment by nesqi** [comment]
What was the verdict?
Did the dcd7c63d09e12fc3e2a9ca80b1422e4bcdd2740f patch fix this problem?

**Comment by speakmore** [comment]
Which release version will include this fix (dcd7c63)?
